chia/util/action_scope.py

```diff
@@ -0,0 +1,174 @@
+from typing import AsyncIterator, Awaitable, Callable, Dict, Generic, List, Optional, Protocol, Type, TypeVar
+
+import aiosqlite
+
+from chia.util.db_wrapper import DBWrapper2, execute_fetchone
+
+
+class ResourceManager(Protocol):
+    @classmethod
+    @contextlib.asynccontextmanager
+    async def managed(cls, initial_resource: SideEffects) -> AsyncIterator[ResourceManager]:  # pragma: no cover
+        # We have to put this yield here for mypy to recognize the function as a generator
+        yield  # type: ignore[misc]
+
+    @contextlib.asynccontextmanager
+    async def use(self) -> AsyncIterator[None]:  # pragma: no cover
+        # We have to put this yield here for mypy to recognize the function as a generator
+        yield
+
+    async def get_resource(self, resource_type: Type[_T_SideEffects]) -> _T_SideEffects: ...
+
+    async def save_resource(self, resource: SideEffects) -> None: ...
+
+    async def get_memos(self) -> Dict[bytes, bytes]: ...
+
+    async def save_memos(self, memos: Dict[bytes, bytes]) -> None: ...
+
+
+@dataclass
+class SQLiteResourceManager:
+
+    _db: DBWrapper2
+    _active_writer: Optional[aiosqlite.Connection] = field(init=False, default=None)
+
+    def get_active_writer(self) -> aiosqlite.Connection:
+        if self._active_writer is None:
+            raise RuntimeError("Can only access resources while under `use()` context manager")
+
+        return self._active_writer
+
+    @classmethod
+    @contextlib.asynccontextmanager
+    async def managed(cls, initial_resource: SideEffects) -> AsyncIterator[ResourceManager]:
+        async with DBWrapper2.managed(":memory:", reader_count=0) as db:
+            self = cls(db)
+            async with self._db.writer() as conn:
+                await conn.execute("CREATE TABLE memos(" " key blob," " value blob" ")")
+                await conn.execute("CREATE TABLE side_effects(" " total blob" ")")
+                await conn.execute(
+                    "INSERT INTO side_effects VALUES(?)",
+                    (bytes(initial_resource),),
+                )
+            yield self
+
+    @contextlib.asynccontextmanager
+    async def use(self) -> AsyncIterator[None]:
+        async with self._db.writer() as conn:
+            self._active_writer = conn
+            try:
+                yield
+            finally:
+                self._active_writer = None
+
+    async def get_resource(self, resource_type: Type[_T_SideEffects]) -> _T_SideEffects:
+        row = await execute_fetchone(self.get_active_writer(), "SELECT total FROM side_effects")
+        assert row is not None
+        side_effects = resource_type.from_bytes(row[0])
+        return side_effects
+
+    async def save_resource(self, resource: SideEffects) -> None:
+        await self.get_active_writer().execute("DELETE FROM side_effects")
+        await self.get_active_writer().execute(
+            "INSERT INTO side_effects VALUES(?)",
+            (bytes(resource),),
+        )
```

Comment:
Maybe worth using the upsert alternative to make this a single query?  Though sure, theoretical optimization and all...  If interested, you or I will have to lookup the form.

Reply:
Not sure what you mean so... no? :)

Reply:
https://www.sqlite.org/lang_upsert.html

Reply:
Changed to just always use UPDATE: 94b2d65f53